droidlet/tools/hitl/nsp_retrain_infra.py

```diff
@@ -0,0 +1,259 @@
+        os.chdir(model_out)
+        for i in range(NUM_MODEL_OUTPUT_POLLS):
+            if ((datetime.now() - now).seconds > NSP_RETRAIN_TIMEOUT):
+                raise TimeoutError(f"NSP Retrain child process timed out after {NSP_RETRAIN_TIMEOUT} seconds")
+            if self.slurm_jobs_finished(job_ids):
+                accs = self.get_accs()
+                self.copy_best_model(accs)
+                break
+            time.sleep(MODEL_OUTPUT_POLL_TIME)
+
+        # Save the best model in S3 bucket and close out
+        upload_key = batch_id + "/best_model/best_model.pth" 
+        s3.upload_file("best_model.pth", 'droidlet-hitl', upload_key)
+
+        self.set_finished(True)
+        logging.info(f"NSP Retraining Job finished")
+        return
+
+
+class NSPNewDataListener(JobListener):
+    def __init__(self, batch_id):
+        super(NSPNewDataListener, self).__init__()
+        self.batch_id = batch_id
+        self.new_data_found = False
+
+    def run(self, runner):
+        logging.info(f"NSP New Data Listener running")
+
+        while not self.check_is_finished():
+            time.sleep(LISTENER_SLEEP_TIME)
+            try:
+                prefix = str(self.batch_id) + '/'
+                new_data_key = s3.list_objects_v2(Bucket='droidlet-hitl', Prefix=prefix, Delimiter='/')['Contents'][1]['Key']
+                self.new_data_found = True
+            except KeyError:
+                logging.info(f"New data not yet detected...")
+                continue
+
+            # Search for new data
+            if (self.new_data_found):
+                logging.info(f"NSP Listener has found new data")
+
+                # Initialize retraining job
+                nsp_rt = NSPRetrainingJob()
+                nsp_rt.batch_id =  self.batch_id # Pass the batch_id to the data generator
+                # self.add_parent_jobs(nsp_rt)
+                runner.register_data_generators([nsp_rt])
+
+                logging.info(f"NSP data gen job registered, listener return")
+                self.set_finished(True)
+                return
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--droidlet_dir", default="/private/home/aszlam/fairinternal/droidlet/")
+    parser.add_argument("--full_data_dir", default="agents/craftassist/datasets/full_data/")
+    parser.add_argument("--sweep_runner_dir", default="/checkpoint/aszlam/nsp_cl_scripts/")
+    parser.add_argument("--sweep_config_folder", default="/checkpoint/aszlam/nsp/sweeps/scripts/configs/auto_sweep_configs/")
+    parser.add_argument("--sweep_scripts_output_dir", default="/checkpoint/aszlam/nsp/sweeps/scripts/")
+    parser.add_argument("--sweep_name", default="auto")
+    parser.add_argument("--output_dir", default="/checkpoint/aszlam/nsp/sweeps/job_output/")
+    parser.add_argument("--checkpoint_dir", default="/checkpoint/aszlam/nsp/")
+    opts = parser.parse_args()
```

Comment:
pass opts as an argument to the classes that needs it. try not to refer it as a global variable